File: app/src/main/java/com/pigmal/lab/lifecycle/MyApplication.java
package com.pigmal.lab.lifecycle;

import android.app.Application;
import android.arch.lifecycle.GenericLifecycleObserver;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.ProcessLifecycleOwner;
import android.util.Log;

public class MyApplication extends Application {
    static final String TAG = "EX4";
    @Override
    public void onCreate() {
        super.onCreate();
        ProcessLifecycleOwner.get().getLifecycle().addObserver(new GenericLifecycleObserver() {
            @Override
            public void onStateChanged(LifecycleOwner source, Lifecycle.Event event) {
                Log.v(TAG, event.name());
            }
        });
    }
}

File: app/src/main/java/com/pigmal/lab/lifecycle/MyApplication.kt
package com.pigmal.lab.lifecycle

import android.app.Application
import android.arch.lifecycle.GenericLifecycleObserver
import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleOwner
import android.arch.lifecycle.ProcessLifecycleOwner
import android.content.res.Configuration
import android.util.Log

class MyApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        trace()

        ProcessLifecycleOwner.get().lifecycle.addObserver(GenericLifecycleObserver { _, event ->
            trace()
        })
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
        trace()
    }

    override fun onTerminate() {
        super.onTerminate()
        trace()
    }
}
